### ISSUE TITLE
Blender 4.1.1 py3dna DLL load failed 

### ISSUE THREAD

**Issue by Igameart** [EpicGames/MetaHuman-DNA-Calibration#55]
Please don't crucify me for posting this, but I'm trying to use the python cmd version of dna/dnacalib in blender but I need to know if it's even possible in the first place. I cannot afford maya, and the MetaPipe and MetaTransfer tools are both expensive so I am trying to create a low cost alternative for blender.

My addon contains a folder called dnacalib, which contains the x64 release builds of dnacalib, which I built myself in python 3.10.

I have the following lines at the top of my addon script:
```
current_dir = os.path.dirname(os.path.abspath(__file__))
dnacalib_dir = os.path.join(current_dir, 'dnacalib')

sys.path.append(dnacalib_dir)
try:
    import _py3dnacalib
except ImportError as e:
    print(f"Failed to import _py3dnacalib: {e}")

try:
    from dna import DataLayer_All, FileStream, Status, BinaryStreamReader, JSONStreamWriter
except ImportError as e:
    print(f"Failed to import from dna: {e}")
```

Attempting to import these gives me the following error 100% of the time:
```
Failed to import _py3dnacalib: DLL load failed while importing _py3dnacalib: The specified module could not be found.
Failed to import from dna: DLL load failed while importing _py3dna: The specified module could not be found.
```

I am new to python, new to writing blender addons, and new to working with MetaHuman DNA. But already, I've got an addon that can retarget a rig and geometries from a base metahuman mesh, to a modified mesh with the same topology.
![image](https://github.com/user-attachments/assets/fa26e985-97dd-42e2-8cba-cd9f2f8337bc)

**Comment by Igameart** [comment]
Okay, new development. I realized I may have compiled with the wrong python version, which I did. I recompiled it with version 3.11.7, which matches my blender environment, now I am getting a new error:

`NameError: file /dna/lib/window/dna.py line 359: name 'FileStream_create' is not defined`

**Comment by n0phx** [comment]
Please try again with the latest release, the above mentioned problem with `FileStream_create` was due to the used Swig version for the build. Earlier only Swig 4.0.x was supported, now the latest release also works on 4.1.x. In any case, this should no longer happen.

**Comment by Igameart** [comment]
I'm going to try this shortly. My current solution requires a custom build of dnacalib, and I don't know if I'm allowed to share that with my tool once it's released, so if this fixes it that would be awesome.

**Comment by n0phx** [comment]
Glad to hear that!

**Comment by Igameart** [comment]
Yep! The addon now clones the repo, builds the source with cmake, then copies each compiled element into a folder the addon can access. So I no longer have to worry about sharing custom builds and potentially violating the terms of service. 

Thanks for the fix!

**Comment by kalebzaki4** [comment]
No worries! It's great that you're exploring options for using DNA/DNAcalib in Blender. While the primary development of DNA/DNAcalib has been focused on Maya, there may be ways to use it in Blender, but there are some important considerations and steps you need to take to troubleshoot your current setup.

### Possible Steps to Resolve the Issue

1. **Verify Compatibility**: Ensure that the version of DNA/DNAcalib you built is compatible with Blender’s Python version. Blender comes with its own Python interpreter, which may differ from the one you're using for development.

2. **Build Configuration**: When building the DLLs for `_py3dnacalib`, make sure they are built with the correct configurations for the version of Python that Blender uses. For instance, if Blender is using Python 3.9, you need to ensure your build matches that version, not 3.10.

3. **Check Architecture**: Ensure that the architecture (x64) of your build matches Blender's architecture. If you're running a 64-bit version of Blender, the DLLs should also be 64-bit.

4. **Dependency Requirements**: Check if `_py3dnacalib` has any additional dependencies (DLLs) that also need to be present in the same directory or in your system's PATH. Use a tool like **Dependency Walker** to check for any missing dependencies.

5. **Correct Directory Structure**: Ensure that your `dnacalib` folder structure is correct and that all required files (including `_py3dnacalib.pyd`, other modules, and dependencies) are in the expected locations.

6. **Using the Right Import Path**: Since you're modifying `sys.path`, confirm that it correctly points to the directory containing the compiled DLLs. You might want to add a print statement to verify the path being added to `sys.path`.

7. **Testing in Blender's Python Console**: Instead of running your addon directly, you could open Blender and test the imports in the Python console to see if the errors persist. This can help you narrow down if the issue is with the addon setup or the module itself.

### Example Directory Structure:
Your directory might look something like this:

```
your_addon/
│
├── __init__.py
├── your_addon_script.py
└── dnacalib/
    ├── _py3dnacalib.pyd
    ├── other_required_files.py
    └── ...
```

### Example Code Snippet:
Make sure your import code looks like this:

```python
import os
import sys

# Get the directory of the current file
current_dir = os.path.dirname(os.path.abspath(__file__))
dnacalib_dir = os.path.join(current_dir, 'dnacalib')

# Add the dnacalib directory to sys.path
sys.path.append(dnacalib_dir)

# Try importing the modules
try:
    import _py3dnacalib
    print("Successfully imported _py3dnacalib")
except ImportError as e:
    print(f"Failed to import _py3dnacalib: {e}")

try:
    from dna import DataLayer_All, FileStream, Status, BinaryStreamReader, JSONStreamWriter
    print("Successfully imported from dna")
except ImportError as e:
    print(f"Failed to import from dna: {e}")
```

### Conclusion:
While using DNA/DNAcalib directly in Blender may have some hurdles, with the right environment setup and dependencies in place, it is feasible. If you continue to encounter issues, consider reaching out to the Blender development community or forums, where users may have tackled similar challenges.

Feel free to ask if you have more questions or need further assistance! Good luck with your addon development!

**Comment by booomji** [comment]
Hey do you have an getting started tutorial on this ?
If I clone your repo how can I get the dna viewer to show up in blender ? Cant wait to get this working.

Cheers,
b

**Comment by Igameart** [comment]
The dna viewer won't work in blender out of the box. You've got to write a mesh building script to innterpret the dna as a blender ready mesh. I'm writing one of my own that's partially complete, but it's a lot of work so I'm not releasing it for free lol
________________________________

**Comment by booomji** [comment]
Hey totally understand . Good luck and write back when it's functional.
Cheers,
b

**Comment by kalebzaki4** [comment]
omg
